src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Windows/HelpersWindows.cs

```diff
@@ -0,0 +1,367 @@
+        {
+            int cbData = 0;
+            if (!Interop.Crypt32.CryptMsgGetParam(hCryptMsg, paramType, index, null, ref cbData))
+                throw Marshal.GetLastWin32Error().ToCryptographicException();
+
+            byte[] pvData = new byte[cbData];
+            if (!Interop.Crypt32.CryptMsgGetParam(hCryptMsg, paramType, index, pvData, ref cbData))
+                throw Marshal.GetLastWin32Error().ToCryptographicException();
+
+            return pvData.Resize(cbData);
+        }
+
+        // Used for binary blobs with internal pointers.
+        public static SafeHandle GetMsgParamAsMemory(this SafeCryptMsgHandle hCryptMsg, CryptMsgParamType paramType, int index = 0)
+        {
+            int cbData = 0;
+            if (!Interop.Crypt32.CryptMsgGetParam(hCryptMsg, paramType, index, null, ref cbData))
+                throw Marshal.GetLastWin32Error().ToCryptographicException();
+
+            SafeHandle pvData = SafeHeapAllocHandle.Alloc(cbData);
+            if (!Interop.Crypt32.CryptMsgGetParam(hCryptMsg, paramType, index, pvData.DangerousGetHandle(), ref cbData))
+                throw Marshal.GetLastWin32Error().ToCryptographicException();
+
+            return pvData;
+        }
+
+        public static byte[] ToByteArray(this DATA_BLOB blob)
+        {
+            if (blob.cbData == 0)
+                return Array.Empty<byte>();
+
+            int length = (int)(blob.cbData);
+            byte[] data = new byte[length];
+            Marshal.Copy(blob.pbData, data, 0, length);
+            return data;
+        }
+
+        public static CryptMsgType GetMessageType(this SafeCryptMsgHandle hCryptMsg)
+        {
+            int cbData = sizeof(CryptMsgType);
+            CryptMsgType cryptMsgType;
+            if (!Interop.Crypt32.CryptMsgGetParam(hCryptMsg, CryptMsgParamType.CMSG_TYPE_PARAM, 0, out cryptMsgType, ref cbData))
+                throw Marshal.GetLastWin32Error().ToCryptographicException();
+            return cryptMsgType;
+        }
+
+        public static int GetVersion(this SafeCryptMsgHandle hCryptMsg)
+        {
+            int cbData = sizeof(int);
+            int version;
+            if (!Interop.Crypt32.CryptMsgGetParam(hCryptMsg, CryptMsgParamType.CMSG_VERSION_PARAM, 0, out version, ref cbData))
+                throw Marshal.GetLastWin32Error().ToCryptographicException();
+            return version;
+        }
+
+        /// <summary>
+        /// Returns the inner content of the CMS.
+        /// 
+        /// Special case: If the CMS is an enveloped CMS that has been decrypted and the inner content type is Oids.Pkcs7Data, the returned
+        /// content bytes are the decoded octet bytes, rather than the encoding of those bytes. This is a documented convenience behavior of 
+        /// CryptMsgGetParam(CMSG_CONTENT_PARAM) that apparently got baked into the behavior of the managed EnvelopedCms class.
```

Comment:
I haven't made it far enough yet to know the answer, but: Do we have a test for this, so that when we eventually make the Unix port that it ensures this convenience?